File: tictactoe.cpp
#include <iostream>

void printgrid(char grid[3][3]);
bool checkwin(char, char grid[3][3]);
void newgame(char (&grid)[3][3]);

int Xwins = 0;
int Owins = 0;
char turn = 'X';
int movecount = 0;

using namespace std;

int main()
{
  char grid[3][3];
  bool running = true;
  char coords[3];

  newgame(grid);
  while(running == true)
    { 
      cout << turn << "'s turn. Enter coords (letter first):" << endl;
      cin.get(coords, 3);
      cin.get();
      bool validcoords = false;
      while(validcoords == false)
	{
	  //if coords are valid
	  if((coords[0] == 'a'||coords[0] == 'b' || coords[0] == 'c') && (coords[1] == '1' ||
								    coords[1] == '2' || coords[1] == '3'))
	    {
	      //if the coords are empty
	      if(grid[int(coords[0]) - 97][(int)coords[1] - 49] == ' ')
		{
		  movecount++;
		  if(turn == 'X')
		    { 
		      grid[int(coords[0]) - 97][(int)coords[1] - 49] = 'X';
		      turn = 'O';
		      validcoords = true;
		    }
 		  else
		    {
		      
		      grid[int(coords[0]) - 97][(int)coords[1] - 49] = 'O';
		      turn = 'X';
		      validcoords = true;
		    }
		} 
	      //if coords have already been played
	      else
		{
		  cout << "These coords have already been played. Try again:" << endl;
		  cin.get(coords, 3);
		  cin.get();
		}
	    }
	  else
	    {
	      cout << "Invalid coords. Try again:" << endl;
	      cin.get(coords, 3);
	      cin.get();
	    }
	}

      printgrid(grid);
      if(checkwin('X', grid) == true)//x has won
	{
	  cout << "Player X has won!" << endl;
	  Xwins++;
	  newgame(grid);
	}
      else if(checkwin('O', grid) == true)//o has won
	{
	  cout << "Player O has won!" << endl;
	  Owins++;
	  newgame(grid);
	}
      else if(movecount == 9) //if board is full and no one has won, it's a tie 
	{
	  cout << "There has been a tie" << endl;
	  newgame(grid);
	}
    }
}

//prints visual tictactoe grid
void printgrid(char grid[3][3])
{
  cout << endl;
  cout << "    1   2   3" << endl;
  cout << endl;
  int letter = 97;
  for(int y = 0; y < 3; y++)
    {
      cout << (char)letter;
      letter++;
      for(int x = 0; x < 3; x++)
        {
	  cout << "   " << grid[y][x];
	}
      cout << endl;
      cout << endl;
    }
}

  //checks if a player has gotten 3 in a row
  bool checkwin(char player, char grid[3][3])
  {
    if(grid[0][0] == player && grid[0][1] == player && grid[0][2] == player)
      {
	return true;
      }
    if(grid[1][0] == player && grid[1][1] == player && grid[1][2] == player)
      {
	return true;
      }
    if(grid[2][0] == player && grid[2][1] == player && grid[2][2] == player)
      {
	return true;
      }
    if(grid[0][0] == player && grid[1][0] == player && grid[2][0] == player)
      {
	return true;
      }
    if(grid[0][1] == player && grid[1][1] == player && grid[2][1] == player)
      {
	return true;
      }
    if(grid[0][2] == player && grid[1][2] == player && grid[2][2] == player)
      {
	return true;
      }
    if(grid[0][0] == player && grid[1][1] == player && grid[2][2] == player)
      {
	return true;
      }
    if(grid[0][2] == player && grid[1][1] == player && grid[2][0] == player)
      {
	return true;
      }
    return false;
  }

//makes grid empty and starts a new game
void newgame(char (&grid)[3][3])
{
  cout << endl;
  cout << "X total wins: " << Xwins << endl;
  cout << "O total wins: " << Owins << endl;
  for(int y = 0; y < 3; y++)
    {
      for(int x = 0; x < 3; x++)
	{
	  grid[y][x] = ' ';
	}
    }
  movecount = 0;
  turn = 'X';
  printgrid(grid);
}

File: TicTacToe.java
/*
 * TicTacToe - players choose their coordinates to place their piece on the board, first to get 3 in a row wins
 * Author: Emily MacPherson
 * Date:10/1/20
 */

import java.util.Scanner;

public class TicTacToe 
{
	int[][] board = new int[3][3];
	final int BLANK = 0;
	final int X_MOVE = 1;
	final int O_MOVE = 2;
	final int X_TURN = 0;
	final int O_TURN = 1;
	int turn = X_TURN;
	int xWins = 0;
	int oWins = 0;
	int ties = 0;
	
	Scanner scanner;
	String input = "";
	
	public static void main(String[] args) 
	{
		new TicTacToe();
	}
	
	//does basic game functions
	public TicTacToe()
	{
		scanner = new Scanner(System.in);
		boolean stillPlaying = true;
		while(stillPlaying == true)
		{
			while(checkWin(X_MOVE) == false && checkWin(O_MOVE) == false && checkTie() == false)
			{
				printBoard();
				//checks if player coordinates are valid
				if(turn == X_TURN)
				{
					System.out.println("Player X, enter your coordinates.");
				}
				else
				{
					System.out.println("Player O, enter your coordinates.");
				}
				input = scanner.nextLine();
				if(input.length() != 2)
				{
					System.out.println("Enter a letter followed by a number");
				}
				else if(input.charAt(0) != 'a' &&
						input.charAt(0) != 'b' && 
						input.charAt(0) != 'c')
				{
					System.out.println("Row must be an a, b, or c.");
				}
				else if(input.charAt(1) != '1' &&
						input.charAt(1) != '2' && 
						input.charAt(1) != '3')
				{
					System.out.println("Column must be an 1, 2, or 3.");
				}
				else
				{
					//places X or O on board
					int row = input.charAt(0) - 'a';
					int column = input.charAt(1) - '1';
					if(board[row][column] == BLANK)
					{
						if(turn == X_TURN)
						{
							board[row][column] = X_MOVE;
							turn = O_TURN;
						}
						else
						{
							board[row][column] = O_MOVE;
							turn = X_TURN;
						}
					}
					else
					{
						System.out.println("There's already a piece there, fool");
					}
				}
			}
			//checks for win and restarts if the player wants
			if(checkWin(X_MOVE) == true)
			{
				System.out.println("X wins!");
				xWins += 1;
				printBoard();
			}
			else if(checkWin(O_MOVE) == true)
			{
				System.out.println("O wins!");
				oWins += 1;
				printBoard();
			}
			System.out.println("X has won " + xWins + " times, O has won " + oWins + " times, and there have been " +
			ties + " ties.");
			System.out.println("Would you like to play again?(yes/no)");
			input = scanner.nextLine();
			if(input.equals("no"))
			{
				System.out.println("Fine, be like that. Have a good day.");
				stillPlaying = false;
			}
			else
			{
				for (int row = 0; row < board.length; row++) 
				{
					for (int column = 0; column < board[0].length; column++) 
					{
						board[row][column] = BLANK;
						turn = X_TURN;
					}
				}
			}
		}
	}
	
	//prints the game board
	public void printBoard()
	{
		System.out.println(" \t1\t2\t3");
		for (int row = 0; row < board.length; row++) 
		{
			String output = (char)('a' + row) + "\t";
			for (int column = 0; column < board[0].length; column++) 
			{
				if(board[row][column] == BLANK)
				{
					output += " \t";
				}
				else if(board[row][column] == X_MOVE)
				{
					output += "X\t";
				}
				else if(board[row][column] == O_MOVE)
				{
					output += "O\t";
				}
			}
			System.out.println(output);
		}
	}
	
	//checks if a play has gotten 3 in a row of their piece
	public boolean checkWin(int player)
	{
		if(board[0][0] == player && board[0][1] == player && board[0][2] == player)
		{
			return true;
		}
		if(board[1][0] == player && board[1][1] == player && board[1][2] == player)
		{
			return true;
		}
		if(board[2][0] == player && board[2][1] == player && board[2][2] == player)
		{
			return true;
		}
		if(board[0][0] == player && board[1][0] == player && board[2][0] == player)
		{
			return true;
		}
		if(board[0][1] == player && board[1][1] == player && board[2][1] == player)
		{
			return true;
		}
		if(board[0][2] == player && board[1][2] == player && board[2][2] == player)
		{
			return true;
		}
		if(board[0][0] == player && board[1][1] == player && board[2][2] == player)
		{
			return true;
		}
		if(board[0][2] == player && board[1][1] == player && board[2][0] == player)
		{
			return true;
		}
		return false;
	}
	
	//checks if any of the board spaces are empty. If none are empty and there hasn't been a win, it's a tie.
	public boolean checkTie()
	{
		for (int row = 0; row < board.length; row++) 
		{
			for(int column = 0; column < board[0].length; column++)
			{
				if(board[row][column] == BLANK)
				{
					return false;
				}
			}
		}
		ties += 1;
		printBoard();
		System.out.println("You have tied.");
		return true;
	}
}
